File: tags/symmetric-build-3.5.19+b.1/symmetric-build-2.5.1/symmetric-ddl/src/main/java/org/jumpmind/symmetric/ddl/platform/sqlite/SqLitePlatform.java
package org.jumpmind.symmetric.ddl.platform.sqlite;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import java.sql.Types;

import org.jumpmind.symmetric.ddl.Platform;
import org.jumpmind.symmetric.ddl.PlatformInfo;
import org.jumpmind.symmetric.ddl.platform.PlatformImplBase;

/*
 * The platform implementation for the SQLite database.
 * 
 * @version $Revision: 231306 $
 */
public class SqLitePlatform extends PlatformImplBase implements Platform {

    /* Database name of this platform. */
    public static final String DATABASENAME = "SQLite3";
    
    /* The standard H2 driver. */
    public static final String JDBC_DRIVER = "org.sqlite.JDBC";
    
    /* The sub protocol used by the H2 driver. */
  // public static final String JDBC_SUBPROTOCOL = "h2";

    /* Below here copied from H2.  May still need tweaking */
    
    /*
     * Creates a new instance of the H2 platform.
     */
    public SqLitePlatform() {
        PlatformInfo info = getPlatformInfo();

        info.setNonPKIdentityColumnsSupported(false);
        info.setIdentityOverrideAllowed(false);
        info.setSystemForeignKeyIndicesAlwaysNonUnique(true);
        info.setNullAsDefaultValueRequired(false);
        info.addNativeTypeMapping(Types.ARRAY, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.DISTINCT, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.NULL, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.REF, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.STRUCT, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.DATALINK, "BINARY", Types.BINARY);

        info.addNativeTypeMapping(Types.BIT, "BOOLEAN", Types.BIT);
        info.addNativeTypeMapping(Types.TINYINT, "SMALLINT", Types.TINYINT);
        info.addNativeTypeMapping(Types.SMALLINT, "SMALLINT", Types.SMALLINT);
        info.addNativeTypeMapping(Types.BINARY, "BINARY", Types.BINARY);
        info.addNativeTypeMapping(Types.BLOB, "BLOB", Types.BLOB);
        info.addNativeTypeMapping(Types.CLOB, "CLOB", Types.CLOB);
        info.addNativeTypeMapping(Types.FLOAT, "DOUBLE", Types.DOUBLE);
        info.addNativeTypeMapping(Types.JAVA_OBJECT, "OTHER");

        info.setDefaultSize(Types.CHAR, Integer.MAX_VALUE);
        info.setDefaultSize(Types.VARCHAR, Integer.MAX_VALUE);
        info.setDefaultSize(Types.BINARY, Integer.MAX_VALUE);
        info.setDefaultSize(Types.VARBINARY, Integer.MAX_VALUE);

        setSqlBuilder(new SqLiteBuilder(this));
        setModelReader(new SqLiteModelReader(this));
    }

    /*
     * {@inheritDoc}
     */
    public String getName() {
        return DATABASENAME;
    }

}

File: symmetric-client-clib/src/db/SqlitePlatform.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "db/SqlitePlatform.h"

static int table_exists_cb(void *exists, int argc, char **argv, char **columName) {
    *((int *) exists) = argc > 0;
    return 0;
}

int SymSqlitePlatform_table_exists(SymDatabasePlatform *super, char *tableName) {
    int exists = 0;
    char sql[100];
    SymSqlitePlatform *this = (SymSqlitePlatform *) super;
    snprintf(sql, 100, "pragma table_info('%s')", tableName);
    sqlite3_exec(this->db, sql, &table_exists_cb, &exists, NULL);
    return exists;
}

int SymSqlitePlatform_execute_sql(SymDatabasePlatform *super,
        const char *sql, int (*callback)(void*, int, char **, char **),
        void *arg, char **errorMessage) {
    SymSqlitePlatform *this = (SymSqlitePlatform *) super;
    return sqlite3_exec(this->db, sql, callback, arg, errorMessage);
}

void SymSqlitePlatform_free(void *data) {
    sqlite3_free(data);
}

void SymSqlitePlatform_destroy(SymDatabasePlatform *super) {
    printf("Closing SQLite database\n");
    SymSqlitePlatform *this = (SymSqlitePlatform *) super;
    sqlite3_close(this->db);
    free(this);
}

SymSqlitePlatform * SymSqlitePlatform_new(SymSqlitePlatform *this, SymProperties *properties) {
    if (this == NULL) {
        this = (SymSqlitePlatform *) calloc(1, sizeof(SymSqlitePlatform));
    }
    SymDatabasePlatform_new(&this->super, properties);
    SymDatabasePlatform *super = &this->super;
    super->name = SYM_DATABASE_SQLITE;
    super->version = (char *) sqlite3_libversion();
    super->execute_sql = (void *) &SymSqlitePlatform_execute_sql;
    super->table_exists = (void *) &SymSqlitePlatform_table_exists;
    super->free = (void *) &SymSqlitePlatform_free;
    super->destroy = (void *) &SymSqlitePlatform_destroy;

    printf("The IDatabasePlatform being used is SymSqlitePlatform\n");

    char filename[100];
    char *url = properties->get(properties, SYM_PARAMETER_DB_URL, SYM_DATABASE_SQLITE);
    url += strlen(SYM_DATABASE_SQLITE) + 1;
    strncpy(filename, url, 100);
    printf("Opening SQLite database at %s\n", filename);

    if (sqlite3_open_v2(filename, &this->db,
            SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_URI, NULL)) {
        fprintf(stderr, "Can't open database: %s\n", sqlite3_errmsg(this->db));
        sqlite3_close(this->db);
        return NULL;
    }

    printf("Detected database '%s', version '%s'\n", super->name, super->version);

    return this;
}
